src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java

```diff
@@ -126,68 +112,120 @@ public ReplicateSupplyService(ReplicatesService replicatesService,
+        // Check is only here to prevent
+        // "`Optional.get()` without `isPresent()` warning".
+        // This case should not happen.
+        if (oReplicatesList.isEmpty()) {
+            return false;
+        }
 
-            final List<Replicate> replicates = replicatesService.getReplicates(chainTaskId);
-            final boolean taskNeedsMoreContributions = ConsensusHelper.doesTaskNeedMoreContributionsForConsensus(
-                    chainTaskId,
-                    replicates,
-                    task.getTrust(),
-                    task.getMaxExecutionTime());
-
-            if (taskNeedsMoreContributions) {
-                String enclaveChallenge = smsService.getEnclaveChallenge(chainTaskId, isTeeTask);
-                if (enclaveChallenge.isEmpty()) {
-                    taskService.unlockTaskAccessForNewReplicate(chainTaskId);//avoid dead lock
-                    continue;
-                }
-
-                replicatesService.addNewReplicate(chainTaskId, walletAddress);
-                taskService.unlockTaskAccessForNewReplicate(chainTaskId);
-                workerService.addChainTaskIdToWorker(chainTaskId, walletAddress);
-
-                // generate contribution authorization
-                return Optional.of(signatureService.createAuthorization(
-                        walletAddress, chainTaskId, enclaveChallenge));
-            }
-            taskService.unlockTaskAccessForNewReplicate(chainTaskId);
+        final ReplicatesList replicatesList = oReplicatesList.get();
+
+        final boolean hasWorkerAlreadyParticipated =
+                replicatesService.hasWorkerAlreadyParticipated(replicatesList, walletAddress);
+        if (hasWorkerAlreadyParticipated) {
+            return false;
         }
 
-        return Optional.empty();
+        if (!taskAccessForNewReplicateLock.lockIfPossible(chainTaskId)) {
+            return false;
+        }
+
+        // no need to go further if the consensus is already reached on-chain
+        // the task should be updated since the consensus is reached but it is still in RUNNING status
+        if (taskUpdateManager.isConsensusReached(replicatesList)) {
+            taskAccessForNewReplicateLock.unlock(chainTaskId);
+            taskUpdateManager.publishUpdateTaskRequest(chainTaskId);
+            return false;
+        }
+
+        // Check if task is still in contribution phase with INITIALIZED or RUNNING status
+        Optional<Task> upToDateTask = taskService.getTaskByChainTaskId(chainTaskId);
+        final boolean taskNeedsMoreContributions = ConsensusHelper.doesTaskNeedMoreContributionsForConsensus(
+                chainTaskId,
+                replicatesList.getReplicates(),
+                task.getTrust(),
+                task.getMaxExecutionTime());
+
+        if (upToDateTask.isEmpty()
+                || !TaskStatus.isInContributionPhase(upToDateTask.get().getCurrentStatus())
```

Comment:
I don't think `isInContributionPhase` is still relevant since `isConsensusReached` already checks on-chain status

Reply:
Absolutely right, thanks!

Reply:
See https://github.com/iExecBlockchainComputing/iexec-core/pull/490/commits/36012c0fead204bc2994a77e71b7d8a98d4e2fac.